ast/src/parsed/display.rs

```diff
@@ -564,6 +564,53 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
     format!("{}", list.into_iter().format(", "))
 }
 
+impl<E: Display> Expression<E> {
+    pub fn precedence(&self) -> Option<ExpressionPrecedence> {
+        match self {
+            Expression::BinaryOperation(_, op, _) => Some(op.precedence()),
```

Comment:
I think we should implement this also for unary operation and LambdaExpression, because those also need parentheses.

Reply:
I fixed the issue for unary operations, but couldn't find an example for lambda, can you help me on this?

Reply:
LambdaEperssion (called "closure" in the rust documentation) should have something like 13 - but I'm not sure if this is still true with the offsets. It should be higher than any binary operation.

Tests would be for example

```
let x = 1 + |i| i + 2;
let x = 1 + (|i| i) + 2;
let x = 1 + |i| (i + 2);
let x = (1 + |i| i) + 2;
let x = (1 + |i| (i + 2));
let x = (1 + |i| i + 2);
```

although you should also test some other binary operators, especially those with high precedence numbers.

Reply:
Three points about lambda expression,
1. Expression like "let x = 1 + |i| i + 2;" cannot be parsed. It seems the parser expects a lambad expression always have parentheses.
2. In formatting Lambda Expression always parentheses are added
https://github.com/powdr-labs/powdr/blob/930f24ac4368ab9e74013ec0cf04bf920c70ec4a/ast/src/parsed/display.rs#L652
3. Our implementation of adding precedence to lambda expression causes this extra parenthesis!
```
Diff < left / right > :
<let x = 1 + (|i| i) + 2;
>let x = 1 + ((|i| i)) + 2;
```
To solve 3, I changed LambdaExpression precedence to `None`, then it worked.
@chriseth Let me know wdyt?

Reply:
Sure, if the other examples still work, then this seems to be the solution.